snapcraft/parts/plugins/kernel_plugin.py

```diff
@@ -193,35 +192,35 @@
 }
 
 
-class KernelPluginProperties(plugins.PluginProperties, plugins.PluginModel):
+class KernelPluginProperties(plugins.PluginProperties, frozen=True):
     """The part properties used by the Kernel plugin."""
 
-    kernel_kdefconfig: List[str] = ["defconfig"]
-    kernel_kconfigfile: Optional[str]
-    kernel_kconfigflavour: Optional[str]
-    kernel_kconfigs: Optional[List[str]]
-    kernel_image_target: Any
+    kernel_kdefconfig: list[str] = ["defconfig"]
+    kernel_kconfigfile: str | None = None
+    kernel_kconfigflavour: str | None = None
+    kernel_kconfigs: list[str] | None = None
+    kernel_image_target: Any = None
     kernel_with_firmware: bool = True
-    kernel_device_trees: Optional[List[str]]
+    kernel_device_trees: list[str] | None = None
     kernel_build_efi_image: bool = False
-    kernel_compiler: Optional[str]
-    kernel_compiler_paths: Optional[List[str]]
-    kernel_compiler_parameters: Optional[List[str]]
-    kernel_initrd_modules: Optional[List[str]]
-    kernel_initrd_configured_modules: Optional[List[str]]
+    kernel_compiler: str | None = None
+    kernel_compiler_paths: list[str] | None = None
+    kernel_compiler_parameters: list[str] | None = None
+    kernel_initrd_modules: list[str] | None = None
+    kernel_initrd_configured_modules: list[str] | None = None
     kernel_initrd_stage_firmware: bool = False
-    kernel_initrd_firmware: Optional[List[str]]
-    kernel_initrd_compression: Optional[str]
-    kernel_initrd_compression_options: Optional[List[str]]
-    kernel_initrd_overlay: Optional[str]
-    kernel_initrd_addons: Optional[List[str]]
+    kernel_initrd_firmware: list[str] | None = None
+    kernel_initrd_compression: str | None = None
+    kernel_initrd_compression_options: list[str] | None = None
+    kernel_initrd_overlay: str | None = None
+    kernel_initrd_addons: list[str] | None = None
     kernel_enable_zfs_support: bool = False
     kernel_enable_perf: bool = False
     kernel_add_ppa: bool = True
-    kernel_use_llvm: Union[bool, str] = False
+    kernel_use_llvm: bool | str = False
 
     # part properties required by the plugin
-    @root_validator
+    @pydantic.model_validator(mode="before")
```

Comment:
With my suggestion above, you could get rid of the first half of this validator.

Additionally, this seems much better as an AfterValidator then:

```python
@pydantic.model_validator(mode="after")
def _validate_compression_options_requires_compression(self) -> Self:
    if self.kernel_initrd_compression_options and not self.kernel_initrd_compression:
        raise ValueError(
            "kernel-initrd-compression-options requires also kernel-initrd-compression to be defined."
        )
    return self